### PR TITLE
Added the Tiberian Sun beacon

### PR DESCRIPTION
**Pull request by Mailaender**
While cleaning up the existing `PlaceBeacon` I noticed that it is over-engineered for the artwork used in RA2/TS so I created a new simplified trait.

### CODE REVIEW

File: OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+				beacon.Tick();
+		}
+
+		public IEnumerable<IRenderable> Render(WorldRenderer r)
+		{
+			if (!owner.IsAlliedWith(owner.World.RenderPlayer))
+				yield break;
+
+			var palette = r.Palette(isPlayerPalette ? beaconPalette + owner.InternalName : beaconPalette);
+			foreach (var b in beacon.Render(position, palette))
```
Comment (abcdefg30) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: This will crash when `beacon == null`, I guess?

**Comment by abcdefg30** [comment]
Looks good to me otherwise. :+1: / :white_check_mark:

File: OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs (hunk 1 above)
Comment (Mailaender) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: Fixed.

File: OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+			if (beacon != null)
+				beacon.Tick();
+		}
+
+		public IEnumerable<IRenderable> Render(WorldRenderer r)
+		{
+			if (!owner.IsAlliedWith(owner.World.RenderPlayer))
+				yield break;
+
+			if (beacon == null)
```
Comment (abcdefg30) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: Sorry to interrupt again, but I think we could move this check above the `IsAlliedWith` check to save some perf. ^^

Comment (Mailaender) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: Resorted.

File: OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs [hunk 3]
```diff
@@ -26,9 +26,10 @@ public BeaconGlobal(ScriptContext context) : base(context)
 
 		[Desc("Creates a new beacon that stays for the specified time at the specified WPos. " +
 			"Does not remove player set beacons, nor gets removed by placing them.")]
-		public Beacon New(Player owner, WPos position, int duration = 30 * 25, bool showRadarPings = true, string palettePrefix = "player")
+		public Beacon New(Player owner, WPos position, int duration = 30 * 25, bool showRadarPings = true,
+			string palette = "player", bool isPlayerPalette = true, string beaconCollection = "beacon", string arrowSprite = "arrow", string circleSprite = "circle")
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs]: I don't like the idea of exposing sequence / animation names to lua.

File: mods/ts/rules/nod-support.yaml [hunk 4]
```diff
@@ -296,6 +296,9 @@ NAMISL:
 		DisplayBeacon: False
 		DisplayRadarPing: True
 		BeaconPoster:
+		ClockSprite:
```
Comment (pchote) [inline mods/ts/rules/nod-support.yaml]: Nulling out properties in yaml isn't very nice.  The default should be empty, and the other mods override it with their additions.

File: mods/ts/rules/nod-support.yaml [hunk 5]
```diff
@@ -296,6 +296,9 @@ NAMISL:
 		DisplayBeacon: False
```
Comment (pchote) [inline mods/ts/rules/nod-support.yaml]: This doesn't show a beacon, so the overrides below shouldn't even be needed...?

File: OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+using System;
+using System.Collections.Generic;
+using OpenRA.Effects;
+using OpenRA.Graphics;
+using OpenRA.Scripting;
+
+namespace OpenRA.Mods.TS.Effects
+{
+	public class AnimatedBeacon : IEffect
```
Comment (pchote) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: Why did you add all those extra properties to the other beacon type, if you are then making a new one for TS?

File: mods/ts/rules/nod-support.yaml (hunk 5 above)
Comment (Mailaender) [inline mods/ts/rules/nod-support.yaml]: The linter doesn't care for that boolean and will otherwise complain.

File: OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs (hunk 6 above)
Comment (Mailaender) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: The initial plan was to use the existing beacon code, but it is so hard-coded to our CTF + remapped rings workaround that I created a new simplified one. I kept the un-hardcoding for documentation and lint testing.

File: OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs (hunk 3 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Scripting/Global/BeaconGlobal.cs]: Any idea how else to solve it here?

File: OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs (hunk 6 above)
Comment (pchote) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: This has made a bit of a confusing mess.

File: OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
```
Comment (obrakmann) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: I can't comment on the functionality, but you should use the updated copyright header here. It's also 2016 :)

File: OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs (hunk 6 above)
Comment (Mailaender) [inline OpenRA.Mods.TS/Effects/AnimatedBeaconSprite.cs]: Moved that into https://github.com/Mailaender/OpenRA/tree/lint-beacon now.

**Comment by Mailaender** [comment]
Updated.

File: OpenRA.Mods.TS/Player/PlaceSimpleBeacon.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	public class PlaceSimpleBeaconInfo : ITraitInfo
+	{
+		public readonly int Duration = 30 * 25;
+		public readonly string NotificationType = "Sounds";
+		public readonly string Notification = "Beacon";
+
+		public readonly bool IsPlayerPalette = false;
+		[PaletteReference("IsPlayerPalette")] public readonly string Palette = "effect";
+
+		public readonly string BeaconCollection = "beacon";
```
Comment (reaperrr) [inline OpenRA.Mods.TS/Player/PlaceSimpleBeacon.cs]: @pchote recently mentioned that the term `Collection` should only be used for UI.

I think `BeaconImage` would be better here and more consistent with other traits.

File: OpenRA.Mods.TS/Player/PlaceSimpleBeacon.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+		[SequenceReference("BeaconCollection")] public readonly string BeaconSprite = "idle";
```
Comment (reaperrr) [inline OpenRA.Mods.TS/Player/PlaceSimpleBeacon.cs]: I'd suggest `BeaconSequence` here for similar reasons.

File: OpenRA.Mods.TS/Effects/AnimatedBeacon.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+		readonly Animation beacon;
+
+		public AnimatedBeacon(Player owner, WPos position, int duration, string beaconPalette, bool isPlayerPalette, string beaconCollection, string beaconSprite)
+		{
+			this.owner = owner;
+			this.position = position;
+			this.beaconPalette = beaconPalette;
+			this.isPlayerPalette = isPlayerPalette;
+
+			if (!string.IsNullOrEmpty(beaconSprite))
```
Comment (reaperrr) [inline OpenRA.Mods.TS/Effects/AnimatedBeacon.cs]: `beaconSequence`

File: OpenRA.Mods.TS/Effects/AnimatedBeacon.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+			{
+				beacon = new Animation(owner.World, beaconCollection);
```
Comment (reaperrr) [inline OpenRA.Mods.TS/Effects/AnimatedBeacon.cs]: `beaconImage`

**Comment by reaperrr** [comment]
:+1: after that.

**Comment by Mailaender** [comment]
Renamed for consistency.

File: OpenRA.Mods.TS/Effects/AnimatedBeacon.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+		{
+			if (beacon == null)
+				yield break;
+
+			if (!owner.IsAlliedWith(owner.World.RenderPlayer))
+				yield break;
+
+			var palette = r.Palette(isPlayerPalette ? beaconPalette + owner.InternalName : beaconPalette);
+			foreach (var b in beacon.Render(position, palette))
+				yield return b;
```
Comment (RoosterDragon) [inline OpenRA.Mods.TS/Effects/AnimatedBeacon.cs]: Minor quibble: Can you make these `yield break;` into `return Enumerable.Empty<IRenderable>();`? I believe you can then return `beacon.Render(...)` directly as well.

Comment (Mailaender) [inline OpenRA.Mods.TS/Effects/AnimatedBeacon.cs]: Yes, indeed. Updated.

**Comment by obrakmann** [comment]
:+1:
